packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,343 @@
+/// scope, or the node owned by the scope node managed by this widget.
+///
+/// The selection process of the node to move to is determined by the node
+/// traversal policy specified by the nearest enclosing
+/// [DefaultFocusTraversal] widget.
+///
+/// It provides [onFocusChange] as a way to be notified when the focus is given
+/// to or removed from this widget, and allows specification of a
+/// [focusedDecoration] to be shown when its [child] has focus.
+///
+/// The [onKey] argument allows specification of a key even handler that should
+/// be invoked when this node or one of its children has focus.
+///
+/// To manipulate the focus, use methods on [FocusScopeNode]. For instance,
+/// to move the focus to the next node, call
+/// `Focus.of(context).nextFocus()`.
+class FocusScope extends Focus {
+  /// Creates a widget that manages a [FocusScopeNode]
+  ///
+  /// The [child] argument is required and must not be null.
+  ///
+  /// The [autofocus], and [showDecorations] arguments must not be null.
+  const FocusScope({
+    Key key,
+    FocusNode node,
+    @required Widget child,
+    bool autofocus = false,
+    ValueChanged<bool> onFocusChange,
+    FocusOnKeyCallback onKey,
+    String debugLabel,
+  })  : assert(child != null),
+        assert(autofocus != null),
+        super(
+          key: key,
+          child: child,
+          node: node,
+          autofocus: autofocus,
+          onFocusChange: onFocusChange,
+          onKey: onKey,
+          debugLabel: debugLabel,
+        );
+
+  /// Returns the node of the [FocusScope] that most tightly encloses the given
+  /// [BuildContext].
+  ///
+  /// The [context] argument must not be null.
+  static FocusScopeNode of(BuildContext context) {
+    assert(context != null);
+    final _FocusMarker marker = context.inheritFromWidgetOfExactType(_FocusMarker);
+    return marker?.node?.nearestScope ?? context.owner.focusManager.rootScope;
+  }
+
+  /// Returns the ancestor nodes of the [FocusScope] in the given
+  /// [BuildContext].
+  ///
+  /// The [context] argument must not be null.
+  static List<FocusScopeNode> ancestorsOf(BuildContext context) {
+    assert(context != null);
+    final FocusScopeNode parent = FocusScope.of(context);
+    return parent?.ancestors?.cast<FocusScopeNode>()?.toList() ?? <FocusScopeNode>[];
   }
 
   @override
-  Widget build(BuildContext context) {
-    FocusScope.of(context).reparentScopeIfNeeded(widget.node);
-    return Semantics(
-      explicitChildNodes: true,
-      child: _FocusScopeMarker(
-        node: widget.node,
-        child: widget.child,
-      ),
+  _FocusScopeState createState() => _FocusScopeState();
+}
+
+class _FocusScopeState extends _FocusState {
+  @override
+  FocusScopeNode _createNode() {
+    return FocusScopeNode(
+      debugLabel: widget.debugLabel,
+      context: context,
     );
   }
 }
+
+// The InheritedWidget marker for Focus and FocusScope.
+class _FocusMarker extends InheritedWidget {
```

Comment:
if you make this an `InheritedNotifier` (on the `node`), you can implement a `Focus.at` method that returns a boolean that is just whether or not the node is currently focused (using inheritFrom...), and it'll automatically rebuild anyone who uses it whenever the state changes.

Reply:
Cool. Didn't know about InheritedNotifier.